First draft workshop content (#3)
observability-lab-cse/observability-lab-workshop: 02-deploy-application/README.md

> Goal:
>
> 1. Having the application run a full E2E and understanding the pieces.
>

> > 2. Observe the application with basic "instrumentation," including emitted log lines and possibly some metrics using `kubectl`. Get familiar with out-of-the-box AKS cluster observability (e.g., workloads, pods, etc.). Look at Container Insights and its capabilities.

Reviewer: ```suggestion
> 2. Being able to observe the running application with basic "instrumentation", including emitted log lines and basic metrics. Getting familiar with out-of-the-box AKS cluster observability (e.g., workloads, pods, etc.). Checking out the Application Insights and its capabilities.
```
I'm not sure why do we mention kubectl here?
Maybe I misunderstood what you meant with the Container Insights and I thought maybe it's about AppInsights instead?

## 1. Deploy Application

Now, let's start deploying the individual services in our application. There are three components to our solution:

> - Device API: An API that allows us to create and delete devices.

Reviewer: ```suggestion
- Devices API: A Java application with an API that allows us to create and delete devices.
```

> - Devices State Manager: A service which manages the state of the devices in he storage. I will update the device the values with the most recent one from the Device Data Simulator.

Reviewer: ```suggestion
- Devices State Manager: A service which manages the state of the devices in the storage. It updates the device temperature value with the most recent one generated by the Devices Data Simulator.
```

- Devices Data Simulator: Simulates devices that publish their health and temperature.

> > Note: make sure you have sourced the `.env` file before you continue form here.

Reviewer: ```suggestion
> Note: make sure you have prepared the `.env` file before you continue from here.
```

> > Note: make sure you have sourced the `.env` file before you continue form here.

Reviewer: ```suggestion
> Note: make sure you have sourced the `.env` file before you continue from here.
```

```bash
source .env
```

```sh
az acr login --name "acr$ENV_PROJECT_NAME".azurecr.io
```

Before we start deploying the application, let's set up all the secrets in the cluster with an easy command.

```sh
make deploy_secret_store
```

We will not go into more details on how to handle secrets in a K8s cluster, as this is not the main point of this workshop. But if you are interested, we leverage the Secret Store CSI driver to pull secrets from the Key Vault resource. You can read more about this [here](https://learn.microsoft.com/en-us/azure/aks/csi-secrets-store-driver).

### Deploy: Devices API

> Let's start by deploying the Device API application. The code for this service can be found [here](https://github.com/observability-lab-cse/observability-lab/tree/section/02-deploy-application/sample-application/devices-api). It's a Java Spring Boot REST API that allows you to list, create, update, and delete devices from your device registry.

Reviewer: ```suggestion
Let's start by deploying the Devices API application. The code for this service can be found [here](https://github.com/observability-lab-cse/observability-lab/tree/section/02-deploy-application/sample-application/devices-api). It's a Java Spring Boot REST API that allows you to list, create, update, and delete devices from your device registry.
```

The first step is to build and push the image to the registry.

<!-- TODO: from where to run the below commands-->

> ```sh

Reviewer: cd sample-application/devices-api
docker build -t acr<project-name>.azurecr.io/devices-api:latest .

TAG="v1"

> docker tag "device-api" "acr$ENV_PROJECT_NAME".azurecr.io/device-api"

Reviewer: It seems there is one " too many.
```suggestion
docker tag "device-api" "acr$ENV_PROJECT_NAME.azurecr.io/device-api"
```

> docker push "acr$ENV_PROJECT_NAME".azurecr.io/device-api":"$TAG"

Reviewer: Here also one " too many it seems?
```suggestion
docker push "acr$ENV_PROJECT_NAME.azurecr.io/devices-api":"$TAG"
```

          env:
            - name: AZURE_COSMOS_DB_URI
              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: CosmosDBEndpoint
            - name: AZURE_COSMOS_DB_KEY
              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: CosmosDBKey
            - name: AZURE_COSMOS_DB_NAME
              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: CosmosDBName
          readinessProbe:
            httpGet:
              path: /health
              port: 8080
            periodSeconds: 20
            initialDelaySeconds: 20
            failureThreshold: 15
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "kvprovider"

---

apiVersion: v1
kind: Service
metadata:
  name: devices-api-service
spec:
  type: LoadBalancer
  ports:
  - port: 8080
    targetPort: 8080
  selector:
    app: devices-api

```

</details>

### Deploy: Devices State Manager

> Next on our agenda is the deployment of the Device State Manager service. You can access the source code for this service [here](https://github.com/observability-lab-cse/observability-lab/tree/section/02-deploy-application/sample-application/device-manager/DeviceManager). This .NET application plays a critical role by updating the temperature records of devices in the database as data flows in from each device.

Reviewer: I'm getting a 404 on the link to the code.

> Next on our agenda is the deployment of the Device State Manager service. You can access the source code for this service [here](https://github.com/observability-lab-cse/observability-lab/tree/section/02-deploy-application/sample-application/device-manager/DeviceManager). This .NET application plays a critical role by updating the temperature records of devices in the database as data flows in from each device.

Reviewer: ```suggestion
Next on our agenda is the deployment of the Devices State Manager service. You can access the source code for this service [here](https://github.com/observability-lab-cse/observability-lab/tree/section/02-deploy-application/sample-application/device-manager/DeviceManager). This .NET application plays a critical role by updating the temperature records of devices in the database as data flows in from each device.
```

As with our previous steps, we'll need to build the image and ensure it's pushed to your Azure Container Registry (ACR).

```sh
TAG="v1"

> docker tag "device-state-manager" "acr$ENV_PROJECT_NAME".azurecr.io/device-state-manager"

Reviewer: I think we need build before this line:

```
cd sample-application/devices-state-manager/DevicesStateManager
docker build -t acr<project-name>.azurecr.io/devices-state-manager:$TAG .
```

              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: EventHubConnectionStringListen
            - name: EVENT_HUB_NAME
              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: EventHubName
            - name: STORAGE_CONNECTION_STRING
              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: StorageAccountConnectionString
            - name: BLOB_CONTAINER_NAME
              value: event-hub-data
            - name: DEVICE_API_URL
              value: "http://devices-api-service:8080"
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "kvprovider"
---

apiVersion: v1
kind: Service
metadata:
  name: devices-state-manager-service
spec:
  type: LoadBalancer
  ports:
  - port: 8090
    targetPort: 8090
  selector:
    app: devices-state-manager

```

</details>

### Deploy: Devices Data Simulator

> To generate data from virtual devices for testing purposes, we're demploying the device simulator, as also utilized in the [sample](TODO). This simulator effectively generates temperature data at defined intervals for each virtual device and transmits this data as messages to Event Hub.

Reviewer: ```suggestion
To generate data from virtual devices for testing purposes, we're deploying the device simulator, as also utilized in the [sample](TODO). This simulator effectively generates temperature data at defined intervals for each virtual device and transmits this data as messages to Event Hub.
```

> To generate data from virtual devices for testing purposes, we're demploying the device simulator, as also utilized in the [sample](TODO). This simulator effectively generates temperature data at defined intervals for each virtual device and transmits this data as messages to Event Hub.

Reviewer: Shall we include a link to the Device Simulator page here? I think it would be this: https://learn.microsoft.com/en-us/samples/azure-samples/iot-telemetry-simulator/azure-iot-device-telemetry-simulator/

> Given we need to define how many devices we like to simluate you can either use the below yaml and update the device list.

Reviewer: ```suggestion
Given we need to define how many devices we would like to simulate, you can either use the below yaml and update the device list.
```

Or leverage the bash script we have already created, which needs the Device API to be up and running.

```sh
make deploy-devices-simulator
```

> > NOTE: Keep in mind that everytime you change something in the list of your devices (eg. Create/Delete/Update a device the simulator needs to be restarted)

Reviewer: ```suggestion
> NOTE: Keep in mind that every time you change something in the list of your devices (e.g. Create/Delete/Update a device), the simulator needs to be restarted.
```

          resources:
            limits:
              cpu: 500m
              memory: 256Mi
            requests:
              cpu: 100m
              memory: 128Mi
          env:
            - name: EventHubConnectionString
              valueFrom:
                secretKeyRef:
                  name: application-secrets
                  key: EventHubConnectionStringSend
            - name: DeviceList
              value: "DEVICE_NAMES_PLACEHOLDER" # Specify your device names with formate `<device-1>,<device-2>,..,<device-n>`
            - name: MessageCount
              value: "0" # send unlimited
            - name: Interval
              value: "60000" # each device sends message every 1 minute
            - name: Template
              value: '{"deviceId": "$.DeviceId", "deviceTimestamp": "$.Time", "temp": $.DoubleValue}'
            - name: Variables
              value: '[{"name": "DoubleValue", "randomDouble":true, "min":20.00, "max":28.00}]'
```

## Out of the box observability

AKS, we can already have a look at what Azure will give us out of the box.

<!-- TODO: go into details here -->

- Overview of pods and services
- Live logs (no history)

> - AKS level metrics that are available

Reviewer: I think we discussed that can also mention tracing here (missing some of the components at this point, e.g. the EventHub).
